include/tvm/build_module.h

```diff
@@ -182,6 +183,76 @@ EXPORT runtime::Module build(const Array<LoweredFunc>& funcs,
                              Target* target_host,
                              const BuildConfig& config);
 
+/*!
+* \brief Generic function that can be specialized on a per-target basis.
+*/
+class GenericFunc {
+ public:
+  /*!
+  * \brief Set the generic function implementaiton.
+  * \param value The generic function
+  * \param allow_override If true, this call may override a previously registered function. If
+  * false, an error will be logged if the call would override a previously registered function.
+  * \return reference to self.
+  */
+  TVM_DLL GenericFunc& set_generic_func(const PackedFunc value,
+                                        bool allow_override = false);
+  /*!
+  * \brief Register a specialized function
+  * \param tags The tags for this specialization
+  * \param value The specialized function
+  * \param allow_override If true, this call may override previously registered tags. If false,
+  * an error will be logged if the call would override previously registered tags.
+  * \return reference to self.
+  */
+  TVM_DLL GenericFunc& register_func(const std::vector<std::string>& tags,
+                                     const PackedFunc value,
+                                     bool allow_override = false);
+  /*!
+  * \brief Invoke the relevant function for a given target. The function will be invoked with
+  * the target string prepended to the arguments.
+  * \param target The target to specialize for.
+  * \param args The arguments to pass to the function. The target string will be prepended to
+  * these arguments.
+  * \param ret The return value
+  */
+  TVM_DLL void invoke_func(const tvm::Target& target, TVMArgs args, TVMRetValue* ret) const;
```

Comment:
I think we would prefer hide target as a context, possibly in  Thread Local storage, which can be queried.  The invoke can simply be operator()().